drivers/dma/dma_andes_atcdmac300.c

```diff
@@ -0,0 +1,568 @@
+/*
+ * Copyright (c) 2023 Andes Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <errno.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/dma.h>
+#include <soc.h>
+
+#define DT_DRV_COMPAT andestech_atcdmac300
+
+#define LOG_LEVEL CONFIG_DMA_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dma_andes_atcdmac300);
+
+#define ATCDMAC100_MAX_CHAN	8
+
+#define DMA_ABORT(dev)		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x24)
+#define DMA_INT_STATUS(dev)		\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x30)
+
+#define DMA_CH_OFFSET(ch)	(ch * 0x20)
+#define DMA_CH_CTRL(dev, ch)		\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x40 + DMA_CH_OFFSET(ch))
+#define DMA_CH_TRANSIZE(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x44 + DMA_CH_OFFSET(ch))
+#define DMA_CH_SRC_ADDR_L(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x48 + DMA_CH_OFFSET(ch))
+#define DMA_CH_SRC_ADDR_H(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x4C + DMA_CH_OFFSET(ch))
+#define DMA_CH_DST_ADDR_L(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x50 + DMA_CH_OFFSET(ch))
+#define DMA_CH_DST_ADDR_H(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x54 + DMA_CH_OFFSET(ch))
+#define DMA_CH_LL_PTR_L(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x58 + DMA_CH_OFFSET(ch))
+#define DMA_CH_LL_PTR_H(dev, ch)	\
+		(((struct dma_atcdmac300_cfg *)dev->config)->base + 0x5C + DMA_CH_OFFSET(ch))
+
+/* Source burst size options */
+#define DMA_BSIZE_1		(0)
+#define DMA_BSIZE_2		(1)
+#define DMA_BSIZE_4		(2)
+#define DMA_BSIZE_8		(3)
+#define DMA_BSIZE_16		(4)
+#define DMA_BSIZE_32		(5)
+#define DMA_BSIZE_64		(6)
+#define DMA_BSIZE_128		(7)
+#define DMA_BSIZE_256		(8)
+#define DMA_BSIZE_512		(9)
+#define DMA_BSIZE_1024		(10)
+
+/* Source/Destination transfer width options */
+#define DMA_WIDTH_BYTE		(0)
+#define DMA_WIDTH_HALFWORD	(1)
+#define DMA_WIDTH_WORD		(2)
+#define DMA_WIDTH_DWORD		(3)
+#define DMA_WIDTH_QWORD		(4)
+#define DMA_WIDTH_EWORD		(5)
+
+/* Bus interface index */
+#define DMA_INF_IDX0		(0)
+#define DMA_INF_IDX1		(1)
+
+/* DMA Channel Control Register Definition */
+#define DMA_CH_CTRL_SBINFIDX_POS	(31)
+#define DMA_CH_CTRL_SBINF(n)		((n) << DMA_CH_CTRL_SBINFIDX_POS)
+#define DMA_CH_CTRL_SBINF_MASK		(1 << DMA_CH_CTRL_SBINFIDX_POS)
+#define DMA_CH_CTRL_DBINFIDX_POS	(30)
+#define DMA_CH_CTRL_DBINF(n)		((n) << DMA_CH_CTRL_DBINFIDX_POS)
+#define DMA_CH_CTRL_DBINF_MASK		(1 << DMA_CH_CTRL_DBINFIDX_POS)
+#define DMA_CH_CTRL_PRIORITY_HIGH	(1 << 29)
+#define DMA_CH_CTRL_SBSIZE_POS		(24)
+#define DMA_CH_CTRL_SBSIZE_MASK		(0x0f << DMA_CH_CTRL_SBSIZE_POS)
```

Comment:
FIELD_SET/FIELD_GET/GENMASK macros can be used in place of these custom shifts and masks

Reply:
The suggestion is awesome!
It makes our code more clear, thank a lot.
We have update the code for this.